packages/core/src/html/NodeProcessor.js

```diff
@@ -176,6 +176,142 @@ class NodeProcessor {
+   */
+
+  /**
+   * Applies pending highlighting to the code block.
+   * This looks into each line for highlighting data, and if found,
+   * traverses over the line and applies the highlight.
+   * @param node Root of the code block element, which is the 'pre' node
+   */
+  static _highlightCodeBlock(node) {
+    const codeNode = node.children.find(c => c.name === 'code');
+    if (!codeNode) {
+      return;
+    }
+
+    codeNode.children.forEach((line) => {
+      if (!_.has(line.attribs, 'hl-start') || !_.has(line.attribs, 'hl-end')) {
+        return;
+      }
+
+      const start = parseInt(line.attribs['hl-start'], 10);
+      const end = parseInt(line.attribs['hl-end'], 10);
+
+      this._traverseLinePart(line, start, end);
+
+      delete line.attribs['hl-start'];
+      delete line.attribs['hl-end'];
+    });
+  }
+
+  /**
+   * Traverses a line part and applies highlighting if necessary.
+   * @param node The node of the line part to be traversed
+   * @param hlStart The highlight start position, relative to the start of the line part
+   * @param hlEnd The highlight end position, relative to the start of the line part
+   * @returns {[number, boolean | [number, number]]} An array of two items.
+   */
+  static _traverseLinePart(node, hlStart, hlEnd) {
+    // Return value is an array of two items:
+    // 1. The number of characters traversed
+    // 2. Highlighting data to be used by the node's parent. It can be:
+    //    - true (ask to apply highlighting from parent)
+    //    - false (do not process this node further)
+    //    - array of two numbers (only for text nodes, inform parent to
+    //      highlight the text at specified range)
+
+    if (hlEnd <= 0) {
+      // Highlight end has passed, no need to traverse further
+      return [0, false];
+    }
+
+    if (node.type === 'text') {
+      // Node is a text node. It is not an inherent HTML element of its own,
+      // so to actually highlight this text, we have to ask to apply at its parent.
+
+      const cleanedText = utils.unescapeHtml(node.data);
+      const textLength = cleanedText.length;
+
+      if (hlStart >= textLength) {
+        // Highlight start is not in this text
+        return [textLength, false];
+      }
+
+      if (hlStart <= 0 && hlEnd >= textLength) {
+        // Highlight spans across the entirety of text
+        return [textLength, true];
+      }
+
+      // Partial text highlighting
+      return [textLength, [hlStart, hlEnd]];
+    }
+
+    // The remaining possibility is that node is a tag node.
+    // It has at least one child (to contain the text content).
+    // It may have more children, such as inner tag nodes.
+    let curr = 0;
+    const shouldHighlight = node.children.map((child) => {
+      const [traversed, data] = this._traverseLinePart(
+        child, hlStart - curr, hlEnd - curr,
+      );
+      curr += traversed;
+
+      return data;
+    });
+
+    if (shouldHighlight.every(v => v === true)) {
```

Comment:
Would it be better to have `.every(v => v)` if `shouldHighlight` is an array of booleans?

Reply:
Not all items in `shouldHighlight` array is boolean. It is just an array to collate the return value as described in the beginning of the function, which may be `true`, `false`, or an array of two numbers. I explicitly uses `=== true` and `=== false` to not include the array of two numbers (which is truthy in itself).

In hindsight maybe I should call this just `highlightData`